File: Mixter.Tests/Domain/Core/Messages/NotifyFollowerOfFolloweeMessageTest.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Mixter.Domain;
using Mixter.Domain.Core.Messages;
using Mixter.Domain.Core.Messages.Events;
using Mixter.Domain.Core.Messages.Handlers;
using Mixter.Domain.Core.Subscriptions;
using Mixter.Domain.Core.Subscriptions.Events;
using Mixter.Domain.Identity;
using Mixter.Infrastructure;
using Mixter.Infrastructure.Repositories;
using Mixter.Tests.Infrastructure;
using NFluent;

namespace Mixter.Tests.Domain.Core.Messages
{
    [TestClass]
    public class NotifyFollowerOfFolloweeMessageTest
    {
        private static readonly UserId Followee = new UserId("followee@mixit.fr");

        private NotifyFollowerOfFolloweeMessage _handler;
        private EventPublisherFake _eventPublisher;
        private EventsDatabase _database;
        private FollowersRepository _followersRepository;
        private SubscriptionsesRepository _subscriptionsesRepository;

        [TestInitialize]
        public void Initialize()
        {
            _database = new EventsDatabase();
            _eventPublisher = new EventPublisherFake();
            _followersRepository = new FollowersRepository();
            _subscriptionsesRepository = new SubscriptionsesRepository(_database);
            _handler = new NotifyFollowerOfFolloweeMessage(_followersRepository, _eventPublisher, _database, _subscriptionsesRepository);
        }

        [TestMethod]
        public void WhenMessagePublishedByFolloweeThenRaiseFolloweeMessagePublished()
        {
            var follower = new UserId("follower@mixit.fr");
            AddFollower(follower);
            var messagePublished = new MessagePublished(MessageId.Generate(), Followee, "content");

            _handler.Handle(messagePublished);

            Check.That(_eventPublisher.Events)
                .Contains(new FolloweeMessagePublished(new SubscriptionId(follower, Followee), messagePublished.Id));
        }

        [TestMethod]
        public void WhenReplyMessagePublishedByFolloweeThenRaiseTimelineMessagePublished()
        {
            var follower = new UserId("follower@mixit.fr");
            AddFollower(follower);
            var messagePublished = PublishMessage(new UserId("author@mixit.fr"), "Hello");
            var replyMessagePublished = new ReplyMessagePublished(MessageId.Generate(), Followee, "Hello", messagePublished.Id);

            _handler.Handle(replyMessagePublished);

            Check.That(_eventPublisher.Events)
                .Contains(new TimelineMessagePublished(new TimelineMessageId(follower, replyMessagePublished.ReplyId), Followee, replyMessagePublished.ReplyContent));
        }

        [TestMethod]
        public void WhenMessageRepublishedByFolloweeThenRaiseTimelineMessageRepublished()
        {
            var follower = new UserId("follower@mixit.fr");
            AddFollower(follower);
            var author = new UserId("author@mixit.fr");
            var messagePublished = PublishMessage(author, "Hello");
            var messageRepublished = new MessageRepublished(messagePublished.Id, Followee);

            _handler.Handle(messageRepublished);

            Check.That(_eventPublisher.Events)
                .Contains(new TimelineMessagePublished(new TimelineMessageId(follower, messagePublished.Id), author, messagePublished.Content));
        }

        private MessagePublished PublishMessage(UserId author, string content)
        {
            var messageId = MessageId.Generate();
            var messagePublished = new MessagePublished(messageId, author, content);
            _database.Store(messagePublished);

            return messagePublished;
        }

        private void AddFollower(UserId follower)
        {
            _followersRepository.Save(new FollowerProjection(Followee, follower));
            _database.Store(new UserFollowed(new SubscriptionId(follower, Followee)));
        }
    }
}

File: domain/src/test/java/mixter/domain/core/subscription/handlers/NotifyFollowerOfFolloweeMessageTest.java
package mixter.domain.core.subscription.handlers;

import mixter.domain.DomainTest;
import mixter.domain.SpyEventPublisher;
import mixter.domain.core.message.MessageId;
import mixter.domain.core.message.events.MessagePublished;
import mixter.domain.core.subscription.FakeFollowerRepository;
import mixter.domain.core.subscription.FakeSubscriptionRepository;
import mixter.domain.identity.UserId;
import org.junit.Before;
import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class NotifyFollowerOfFolloweeMessageTest extends DomainTest {

    public String CONTENT = "Content";
    public UserId AUTHOR_ID = new UserId("author@mix-it.fr");
    public UserId FOLLOWER_ID = new UserId("follower@mix-it.fr");
    public MessageId MESSAGE_ID = MessageId.generate();

    private SpyEventPublisher eventPublisher;
    private FakeSubscriptionRepository subscriptionRepository;
    private FakeFollowerRepository followerRepository;

    @Before
    public void setUp() throws Exception {
        eventPublisher = new SpyEventPublisher();
        subscriptionRepository = new FakeSubscriptionRepository();
        followerRepository = new FakeFollowerRepository();
    }

    @Test
    public void APublishedMessageShouldNotifyFollowers() {
        // Given
        MessagePublished messagePublished = new MessagePublished(MESSAGE_ID, CONTENT, AUTHOR_ID);
        subscriptionRepository.add(subscriptionFor(new UserFollowed(new SubscriptionId(FOLLOWER_ID, AUTHOR_ID))));
        followerRepository.saveFollower(AUTHOR_ID, FOLLOWER_ID);

        NotifyFollowerOfFolloweeMessage handler = new NotifyFollowerOfFolloweeMessage(followerRepository, subscriptionRepository, eventPublisher);

        // When
        handler.apply(messagePublished);
        // Then
        FolloweeMessagePublished followeeMessagePublished = new FolloweeMessagePublished(new SubscriptionId(FOLLOWER_ID, AUTHOR_ID), MESSAGE_ID);
        assertThat(eventPublisher.publishedEvents).containsExactly(followeeMessagePublished);
    }
}
